### PR TITLE
Research System

### PR DESCRIPTION
**Pull request by Naman-sopho**
Includes: 
- Construction of a new Laboratory building.
- A new Research Task which can be added by taking a book out of the Laboratory building's bookcase and placing it in the building's pedestal.
- Performing the research task. This requires the blocks mentioned in the recipe to be present in the building's chest.
- Addition of fence blocks around a area after task assignment.

Requires this https://github.com/Terasology/Books/pull/16 to be merged.

### How to test
- Select an area and assign the Laboratory construction task. The area should be resized according to the size of the building and fences be placed around this new area.
- After construction execute the command `give oreonplant0` and place the obtained items in Laboratory chest(required by the research task).
- Take the book out of bookcase(consists of the recipe which would be used for the research) and place it in the Pedestal. Oreon should perform the research task which adds a Cookie Crop block to the chest.

### CODE REVIEW

File: src/main/java/org/terasology/research/ResearchSystem.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        if (inventoryEntity.getParentPrefab().getName().equals(Constants.PEDESTAL_PREFAB)) {
+            for (EntityRef item : inventoryComponent.itemSlots) {
+                DisplayNameComponent nameComponent = item.getComponent(DisplayNameComponent.class);
+
+                // Check if item is research Book
+                if (nameComponent != null && nameComponent.name.equals(Constants.RESEARCH_BOOK_NAME)) {
+                    BookComponent bookComponent = item.getComponent(BookComponent.class);
+                    List<String> textList = bookComponent.pages;
+
+                    String recipePrefabName = "";
```
Comment (mkienenb) [inline src/main/java/org/terasology/research/ResearchSystem.java]: You could move this section into a well-named method to make the code easier to read.

File: src/main/java/org/terasology/research/ResearchSystem.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+        Task researchTask = new ResearchTask(blockNameList, result.getURI().toString());
+        researchTask.requiredBuildingEntityID = laboratory.getId();
+
+        TaskComponent taskComponent = new TaskComponent();
+        taskComponent.assignedTaskType = AssignedTaskType.Research;
+        taskComponent.task = researchTask;
+
+        ConstructedBuildingComponent buildingComponent = laboratory.getComponent(ConstructedBuildingComponent.class);
+        taskComponent.taskRegion = buildingComponent.boundingRegions.get(16);
```
Comment (mkienenb) [inline src/main/java/org/terasology/research/ResearchSystem.java]: magic number -- what does 16 mean? Use a constant.

File: src/main/java/org/terasology/taskSystem/TaskManagementSystem.java [hunk 3]
```diff
@@ -235,6 +250,79 @@ public void setTaskType(String newTaskType, BuildingType buildingType, Region3i
...
+    }
+
+    private void placeFenceBlocks(Region3i region, boolean placeTorch) {
+        int minX = region.minX();
+        int maxX = region.maxX();
+        int minZ = region.minZ();
+        int maxZ = region.maxZ();
+        int y = region.minY();
+
+        Block block = blockManager.getBlock("Fences:fence.0");
```
Comment (mkienenb) [inline src/main/java/org/terasology/taskSystem/TaskManagementSystem.java]: Seems like this should be a constant.

File: src/main/java/org/terasology/taskSystem/TaskManagementSystem.java [hunk 4]
```diff
@@ -235,6 +250,79 @@ public void setTaskType(String newTaskType, BuildingType buildingType, Region3i
...
+        for (int x = minX; x <= maxX; x++) {
+            for (int z = minZ; z <= maxZ; z++) {
+                blockEntityRegistry.setBlockForceUpdateEntity(new Vector3i(x, y + 1, z), block);
+            }
+        }
+
+        // Place torches on corners
+        if (placeTorch) {
+            block = blockManager.getBlock("Core:Torch.TOP");
```
Comment (mkienenb) [inline src/main/java/org/terasology/taskSystem/TaskManagementSystem.java]: Seems like this should be a constant.

File: src/main/java/org/terasology/taskSystem/TaskManagementSystem.java [hunk 5]
```diff
@@ -235,6 +250,79 @@ public void setTaskType(String newTaskType, BuildingType buildingType, Region3i
...
+            blockEntityRegistry.setBlockForceUpdateEntity(new Vector3i(maxX, y + 2, maxZ), block);
+        }
+    }
+
+    private Region3i getBuildingExtents(BuildingType buildingType, Region3i region) {
+        Prefab buildingPrefab;
+
+        switch (buildingType) {
+            case Diner:
+                buildingPrefab = prefabManager.getPrefab("MasterOfOreon:inn");
```
Comment (mkienenb) [inline src/main/java/org/terasology/taskSystem/TaskManagementSystem.java]: Seems like this should be a constant.

File: src/main/java/org/terasology/taskSystem/TaskManagementSystem.java [hunk 6]
```diff
@@ -235,6 +250,79 @@ public void setTaskType(String newTaskType, BuildingType buildingType, Region3i
...
+                break;
+
+            default:
```
Comment (mkienenb) [inline src/main/java/org/terasology/taskSystem/TaskManagementSystem.java]: Seems like this should be a constant.

**Review by mkienenb — CHANGES_REQUESTED** [review]
Looks pretty good.   Just needs a few more constants.

**Comment by Cervator** [comment]
Tested out, can confirm the research system appears to work 👍 Nice work there.

While testing this and the other PR (and occasionally overall) I've noticed a variety of quality issues and smaller quirks that aren't necessarily linked to this PR. I added some as card to the project board, would be nice to get some of those addressed and focus some on stabilizing stuff further :-)

**Pull request merged**